doc/requirements/Requirements.md

```diff
@@ -0,0 +1,704 @@
+implementation implements them. Search for "reference implementation".</nb>
+
+# Introduction
+The PF **MUST** be a library.
+<why>To be reused in different components.</why>
+
+PF instances **MUST NOT** share any mutable data.
+<why>to guarantee that different PF instances will not impact each others.</why>
+
+# Parameters
+
+A PF **MUST** be able to handle parameters.
+<why>because the PF aims to abstract hardware and model it by parameters.</why>
+
+## Definitions
+
+<dl>
+<dt>Hardware</dt>
+<dd>System controlled by the PF. Not necessary material system. This term was
+chosen because:
+
+ - historically the PF reference implementation was used to abstract hardware
+ - the subsystem term would arguably fit best is already used.
+
+(FIXME: choose subsystem instead, it is not used in fact.)
+</dd>
+</dl>
+
+## Value
+
+A parameter **MUST** have a value.
+<why>because a parameter without value would not abstract any hardware.</why>
+
+A PF **MUST** support mutable parameters.
+<why>To control the underlined hardware.</why>
+
+A PF **MAY** support immutable parameters, i.e. parameters which value is determined
+on start then read only.
+<why>To permit hardware read only value reflection.</why>
+<ko>This is not implemented in the PF reference implementation.</ko>
+
+This value **MUST** be gettable and settable (except if immutable).
+<why>A parameter that can not be accessed is of no use.</why>
+
+A value **MUST** be convertible to a string.
+<why>To display a parameter value to the human user for debugging purpose.</why>
+
+A bijective conversion string <-> value **SHOULD** exist.
+<why>To support serialization and deserialization of the parameter values. It is
+not a MUST because a PF implementation could offer serialization to custom --
+not string based -- format.</why>
+
+### Data type
+
+#### Definition
+
+<dl>
+<dt>Data type</dt>
+<dd>
+All parameters have a data type. A data type designates parameter invariants.
+
+A data type is the meaning of the data and the way values of that type can be
+stored.
+</dd>
+</dl>
+
+
+#### Philosophy
+
+A data type defines the value properties:
+
+ - memory layout
+ - value constrains
+
+A value type is mostly used to:
+
+ - pretty display parameter values (not just a as an array of bits)
+ - check for user error when setting it (out of bound, invalid...)
+ - offer a type safe API
+
+#### Requirements
+
+A PF **SHOULD** support the following types.
+If a type is chosen to be supported, it **MUST** respect all MUST clause,
+**SHOULD** respect all SHOULD clause, **MAY** respect all MAY clause of the type.
+<why>All type are not necessary to use the PF. For example any parameter could
+be represented as an array of char (string). But this would not permit to
+check parameter validity (invariants) nor a pretty display of the values.</why>
+
+Implementation **MAY** add another API to access a parameter value.
+<why>For example a C++ implementation may give access to a string as an
+std::string object.</why>
+
```

Comment:
I would expect the requirement to be more end-user oriented, thus ideally, the pfw should be agnostic of user's types